### ISSUE TITLE
Auto-inferred generic parameters

### ISSUE THREAD

**Issue by robkuz** [fsharp/fslang-suggestions#629]
New proposal (by @dsyme)

Allow for auto-inferrable generic parameters expressed like this

```fsharp
Foo<'a, [<Inferred>] 'b, [<Inferred>] 'c when 'a :> Bar<'b,'c>>(v:'a) = class end
```
Then only `Foo<int>` is required at each use site and the other generic parameters are optional.  `Foo<int>` would always be shorthand for `Foo<int, _, _>`

In certain situations where wildcards are not allowed, solving the type parameters must be possible "immediately".  Exactly what that means technically would be TBD.

---

Original: 
Allow for type constraints to be expressed like this

```fsharp
Foo<'a when 'a :> Bar<_,_>>(v:'a) = end class
```
this is not possible at the moment and one has to define the type like this

```fsharp
Foo<'a, 'b, 'c when 'a :> Bar<'b,'c>>(v:'a) = end class
```

## Semantics
This is bad because when writing `type Foo<'a when 'a :> Bar<_,_>>(v:'a)` 
One might really only be interested in knowing about `Bar` as such and not about its generic params. 
Requiring this is mentally taxing and unneeded. 

## Noise
Now what would happen if our generic type should implement 2 interfaces
each having their own set of generic type parameters?

```fsharp
type Foo<'a,'b,'c,'x,'y when 'a :> Bar<'b,'c> and 'a :> Baz<'x, 'y>(v: 'a) = end class
type BarAndBaz<'a, 'b, 'c, 'd>() =
    interface Bar<'a, 'b>
    interface Baz<'c, 'd>>
type ConcreteFoo = Foo<BarAndBaz<int, string, bool, float>,int, string, bool, float>

// or as part of a function defintion
let foo (x:Foo<BarAndBaz<'a, 'b, 'c, 'd>, 'a, 'b, 'c, 'd>)
```

This is simply painful on the eyes. The visual noise is numbing AND we are only 2 levels deep sofar.
Imagine this type

```fsharp
type FooWrapper<'a,'b,'c,'d,'e,'f, 'x,'y when 'a :> Foo<'b,'c,'d,'e,'f> and 'a :> SomeOtherType<'x, 'y>>(a:'a) = end class
```

You don't want to put `FooWrapper` anywhere in your code.

## Inference

Could this all be solved by type inference ? Let's see an example

```fsharp
type BarAndBaz<'a, 'b, 'c, 'd>() =
    interface Bar<'a, 'b>
    interface Baz<'c, 'd>>
type ConcreteFoo = Foo<BarAndBaz<'a, 'b, 'c, 'd>, _, _, _, _>>
```

Okay, this is a bit better. Maybe we could even improve this?

```fsharp
type ConcreteFoo = Foo<BarAndBaz<_, _, _, _>, _, _, _, _>>
```

Hmm, this seems acceptable. But lets look what happens with one more level of indirection

```fsharp
type Wrapper<'a,'b,'c,'d,'e, 'x,'y, 'z 
                when 'a :> Foo<'b,'c,'d,'e,'f> 
                and 'x :> SomeOtherType<'y, 'z>>(a:'a, x: 'x) = end class

//and then

type ConreteWrapper = Wrapper<Foo<BarAndBaz<_, _, _, _>, _, _, _, _>>, ConcreteOtherType<<_,_>, _, _>>>
```

Wow! this is really hard to parse for a human. And while it may be acceptable on a standalone type alias,
just imagine 2 or 3 such times on a function definition. Nobody will use that - and nobody does!

So I don't think that inference will help that much.

## Infectious

However the worst part of the actual design/syntax is that it is higly infectious
```fsharp
type Foo<'a, 'b>() = end class
type Bar<'a, 'b, 'c when 'a :> Foo<'b, 'c>>() = end class
type FooBarWrapper<'a, 'b, 'c, 'd when 'a :> Bar<'b, 'c, 'd>> = end class
type FooBarWrapperWrapper<'a,'b,'c,'d,'e when 'a :> FooBarWrapper<'b,'c,'d,'e>> = end class 
```

Just imagine what happens if Foo suddenly needs a 3rd generic param. 
The changes will ripple thru your code base like a wild fire. And for what reason? `FooBarWrapperWrapper` might not
be interested in that additional param. Just that it wraps a `FooBarWrapper`.

The actual syntax when defining type constraints requires us to expose implementation details which we might not be interested in (or not necessarily so) there for I propose that F# allows to define constraints as wildcards

## Proposal

1) Proposal A: check for number generic type params
```fsharp
Foo<'a when 'a :> Bar<_,_>>(v:'a) = end class
```

2) Proposal B (additive) : Full wild cards
```fsharp
Foo<'a when 'a :> Bar<*>>(v:'a) = end class
```
For this to work the compiler has to check that there are not multiple types with varying numbers of 
generic type paramters within the actual namespace.

This proposal is kind of the invers of 
[Allow for Structural Checking of Generic Parameters within Type Constraints](https://github.com/fsharp/fslang-suggestions/issues/566)

## Pros and Cons 

The advantages of making this adjustment to F# are it allows for more generic code and together with
[prposal 566](https://github.com/fsharp/fslang-suggestions/issues/566) could get us really very far into the realm of higher kinded types (conceptually) without breaking C# compatibility.

The disadvantages of making this adjustment to F# are ... I seriously think none - apart that the implementation of such a feature is always error prone. I am convinced that this makes F# even easier to grasp. Also the fact that this can be expressed as type constraints seems a big plus to me - that means novice programmers can be guided away from this at the beginning.  
And as the old syntax should be available still this is also backwards compatible.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): L?

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [X] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [X] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [X] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [X] This is not a breaking change to the F# language design
* [X] I or my company would be willing to help implement and/or test/or pay for this

**Comment by realvictorprm** [comment]
My first impression is that I like it. What do the others think?

**Comment by alfonsogarciacaro** [comment]
An example of something that would benefit from this is the Fable React bindings. The helper to instantiate components from type declaration uses the generic argument to get the type definition:

```fsharp
let inline ofType<'T,'P,'S when 'T :> Component<'P,'S>> (properties: 'P) (children: ReactElement list): ReactElement =
    createElement(typedefof<'T>, props, children)
```

As `React.Component` (from which the type must derive) is generic, we force the user to pass two extra types that can be inferred from an argument ('P) or we don't care about ('S most of the time). So these are always filled with wildcards:

```fsharp
ofType<MyComponent,_,_> { myProps = 5 } []
```

**Comment by abelbraaksma** [comment]
The OP suggests this to become valid:

```f#
Foo<'a when 'a :> Bar<_,_>>(v:'a) = end class
```

But how is this compiled age how is this going to be used? The way I see it, there are still two unknown generic parameters and these will have to end up in the type signature. Sure, the compiler can help with that, but that doesn't remove the burden of having to specify such types on each call (unless using functional inference). 

As such, I'm not sure what the actual benefit is. Could you elaborate with a simple, yet full example that shows the feature in action, both in declaration and use, and the proposed desugaring?

**Comment by robkuz** [comment]
It simply proposes to you could ignore generic params when the containing type is using only methods of the contained type that do not expose said generic params.

    type Bar<_,_> =
          member _.Reduce(): int = ...

would be an example

And as you mentioned it pushes the type params to instantiation or usage time

    let x = Foo(Bar<int, int>) // Foo<Bar<int, int>>

so it should infer those generic type params on its own without me telling that there are generic type params.

However I dont see this anytime soon to be realized so maybe we should simply close this

**Comment by dsyme** [comment]
This is really hard to implement technically - the process of "bringing constraints into existence" is already multi-phase, and adding a new phase that adds new type parameters and fills in wildcards is really difficult.

I do get that in many cases type parameters can be inferred via constraints that effectively specify functional dependencies - this is the real underlying problem here and there should be a way to elide those type inference solutions. Adding an attribute `[<InferThisTypeParameter>]` on the inferrable type parameters seems better rather than trying to reduce what's given at the declaration site

I'm trying to find if there's a suggestion for this - I thought there was but I can't find it.  I'll recycle this issue to track that

**Comment by dsyme** [comment]
@Happypig375 Posssibly

This is actually very close the the approved-in-principle https://github.com/fsharp/fslang-suggestions/issues/261, I will close this and I've merged the adjusted proposal I've written over there
